### PR TITLE
sPt1

### PR DESCRIPTION
**Pull request by MarGabor**
Added a button to the main page that directs to a new page where a grant call text can be entered. Upon clicking "calculate" button, a list of k(here k=10) most similar scientist along with their similarity is displayed.

I cannot connect to http://141.53.34.56, yet. Thus, I decided to leave the page running on localhost for now.

### CODE REVIEW

File: src/templates/displayGrantCallResult.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+    <body>
+        <form>
+            <label id="txt1"> <b>Most similar scientists: </b> <br>
+                {% autoescape false %}
+                {{text1}}
+                {% endautoescape %}
+            </label> <br>
+        </form>
+    </body>
+</html>
```
Comment (MarioStanke) [inline src/templates/displayGrantCallResult.html:11]: newline at end of last line, also in another file

File: src/website.py [hunk 2]
```diff
@@ -63,5 +67,56 @@ def compute_similarity_of_files() -> str:
...
+        grantCallEmbedding = embedding_from_string(grantCall)
+    except RetryError:
+        return render_template("displayGrantCallResult.html"
+                                   , text1="Please enter a text.")
+
+    #get embeddings from publication list file
+    with h5py.File("../data/pub_embeddings.h5", 'r') as hdf:
+        pub_embedding = hdf['publication_embedding'][:]
+        author_ids = hdf['author_ids'][:]
+
```
Comment (MarioStanke) [inline src/website.py:93]: This is outdated.  "embeddings" "ids"
image

**Review by MarioStanke — CHANGES_REQUESTED** [review]

**Comment by MarGabor** [comment]
Fixed stuff. Still runs locally, though.

File: src/templates/grantCallForm.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<html>
+    <body>
+        <form action = "http://localhost:5000/grantcallResult" method = "post">
```
Comment (MarioStanke) [inline src/templates/grantCallForm.html:3]:
```suggestion
        <form action = "/grantcallResult" method = "post">
```
`localhost` would link to the client computer. The LLM functions must be executed on the host, however.

File: src/website.py [hunk 4]
```diff
@@ -63,5 +67,58 @@ def compute_similarity_of_files() -> str:
...
+    #k most similar scientists to display
+    k = 10
+    #get grant call text
+    grantCall = request.form["grantCall"]
+    #calculate embedding for grant call text
+    try:
+        grantCallEmbedding = embedding_from_string(grantCall)
+    except RetryError:
+        return render_template("displayGrantCallResult.html"
+                                   , text1="Please enter a text.")
```
Comment (MarioStanke) [inline src/website.py:87]:
```suggestion
        return render_template("displayGrantCallResult.html",
                                text1="Please enter a text.")
```
Comma on old line.

File: src/website.py [hunk 5]
```diff
@@ -63,5 +67,58 @@ def compute_similarity_of_files() -> str:
...
 
+#navigate to grant call form when button is clicked
+@app.route('/grantcall', methods=['POST', 'GET'])
+def navigateToGrantCallForm() -> str:
+    return render_template("grantCallForm.html")
+
+#calculate similarity and list k most similar scientist upon button click 
+@app.route('/grantcallResult', methods=['POST', 'GET'])
+def calculateGrantCallResult() -> str:
+    #k most similar scientists to display
```
Comment (MarioStanke) [inline src/website.py:78]:
```suggestion
    # k most similar scientists to display
```
Space after hash sign. See coding style.

File: src/website.py [hunk 6]
```diff
@@ -63,5 +67,58 @@ def compute_similarity_of_files() -> str:
...
+    
+    similarityList = []
+    for j in range(0,len(ids)-1):
+        #get associated embedding
+        embedding = embeddings.iloc[j]
+        try:
+            similarity = cosine_similarity(grantCallEmbedding, embedding)
+        except ValueError:
+            return render_template("displayGrantCallResult.html"
+                                   , text1="Please enter a text.")
```
Comment (MarioStanke) [inline src/website.py:104]: Space around binary operator (=). Comma on old line. See coding style.

File: src/website.py [hunk 7]
```diff
@@ -63,5 +67,58 @@ def compute_similarity_of_files() -> str:
...
+        similarityList.append((similarity, int(ids.iat[j,0])))
+    #sort list by similarity
+    similarityList = sorted(similarityList, reverse=True)
+    
+    #output k most similar
+    outputText = ""
+    #get names
+    prof_df = pd.read_table("../data/prof.tbl")
+    for i in range(0,k-1):
```
Comment (MarioStanke) [inline src/website.py:113]:
```suggestion
    for i in range(0, k-1):
```

File: src/website.py [hunk 8]
```diff
@@ -63,5 +67,58 @@ def compute_similarity_of_files() -> str:
...
+    similarityList = []
+    for j in range(0,len(ids)-1):
+        #get associated embedding
+        embedding = embeddings.iloc[j]
+        try:
+            similarity = cosine_similarity(grantCallEmbedding, embedding)
+        except ValueError:
+            return render_template("displayGrantCallResult.html"
+                                   , text1="Please enter a text.")
+        similarityList.append((similarity, int(ids.iat[j,0])))
```
Comment (MarioStanke) [inline src/website.py:105]:
```suggestion
        similarityList.append((similarity, int(ids.iat[j, 0])))
```

File: src/website.py [hunk 9]
```diff
@@ -63,5 +67,58 @@ def compute_similarity_of_files() -> str:
...
+    prof_df = pd.read_table("../data/prof.tbl")
+    for i in range(0,k-1):
+        firstname = prof_df.loc[prof_df.id == similarityList[i][1],
+                                "firstname"].values[0]
+        lastname = prof_df.loc[prof_df.id == similarityList[i][1],
+                                "lastname"].values[0]
+        entry = ("<pre>" + str(firstname) + " " + str(lastname)
+                 + "    " + str(similarityList[i][0]) + "</pre>" + "<br>")
+        outputText += entry
+    return render_template("displayGrantCallResult.html", text1=outputText)
```
Comment (MarioStanke) [inline src/website.py:121]:
```suggestion
    return render_template("displayGrantCallResult.html", text1 = outputText)
```

File: src/website.py (hunk 7 above)
Comment (MarioStanke) [inline src/website.py:113]: This is buggy and will fail if *k* is larger than the number of embedded texts. This functionality has been implemented before in `chunks_for_question.get_k_IDs`

**Review by MarioStanke — CHANGES_REQUESTED** [review]
See individual comments.

**Comment by MarGabor** [comment]
Implemented suggested changes. I could've probably created a new file for the code and encapsulated it more for the purpose of despaghettification. But alas.

File: src/website.py [hunk 10]
```diff
@@ -67,58 +67,66 @@ def compute_similarity_of_files() -> str:
...
 
-#calculate similarity and list k most similar scientist upon button click 
+# calculate similarity and list k most similar scientist upon button click 
 @app.route('/grantcallResult', methods=['POST', 'GET'])
 def calculateGrantCallResult() -> str:
-    #k most similar scientists to display
-    k = 10
-    #get grant call text
+    # k most similar scientists to display
+    k = 109
```
Comment (MarioStanke) [inline src/website.py:79]:
```suggestion
    k = 10
```

File: src/website.py [hunk 11]
```diff
@@ -67,58 +67,66 @@ def compute_similarity_of_files() -> str:
...
     outputText = ""
-    #get names
+    # get names
     prof_df = pd.read_table("../data/prof.tbl")
-    for i in range(0,k-1):
+    # if number of ids is too small, set k to be smaller
+    num_id = prof_df["id"].shape[0]
+    if num_id < k:
+        k = num_id
+    for i in range(0, k-1):
```
Comment (MarioStanke) [inline src/website.py:121]:
```suggestion
    for i in range(k):
```

**Review by MarioStanke — APPROVED** [review]
LGTM
I will reset k = 10 and commit myself.

**Pull request merged**